src/test/java/org/broadinstitute/hellbender/utils/iterators/ShardingIteratorUnitTest.java

```diff
@@ -0,0 +1,167 @@
+        final SimpleInterval paddedInterval = interval.expandWithinContig(pad, TEST_HEADER.getSequenceDictionary());
+        // create a list of reads larger than the interval
+        final ShardingIterator<GATKRead> it = new ShardingIterator<>(
+                // create on a different contig (after)
+                oneBaseReadsUntilPosition(1, 10).iterator(),
+                Collections.singletonList(new ShardBoundary(interval, paddedInterval)),
+                TEST_HEADER.getSequenceDictionary());
+
+        // test the next element
+        testNextElement(it, interval, paddedInterval, 0);
+        // assert that it is exhausted
+        Assert.assertFalse(it.hasNext());
+    }
+
+    @Test(dataProvider = "shardIteratorData")
+    public void testDataOnlyBeforeShards(final int start, final int end, final int pad) {
+        // create a simple interval
+        final SimpleInterval interval = new SimpleInterval(TEST_HEADER.getSequence(1).getSequenceName(), start, end);
+        final SimpleInterval paddedInterval = interval.expandWithinContig(pad, TEST_HEADER.getSequenceDictionary());
+        // create a list of reads larger than the interval
+        final ShardingIterator<GATKRead> it = new ShardingIterator<>(
+                // create on a previous contig
+                oneBaseReadsUntilPosition(0, 10).iterator(),
+                Collections.singletonList(new ShardBoundary(interval, paddedInterval)),
+                TEST_HEADER.getSequenceDictionary());
+
+        // test the next element
+        testNextElement(it, interval, paddedInterval, 0);
+        // assert that it is exhausted
+        Assert.assertFalse(it.hasNext());
+    }
+
+    @Test(dataProvider = "shardIteratorData")
+    public void testNonOverlappingShards(final int start, final int end, final int pad) {
+        // create boundaries with intervals
+        final List<GATKRead> reads = new ArrayList<>();
+        final List<ShardBoundary> shards = new ArrayList<>();
+        for (int i = 0; i < 2; i++) {
+            final SimpleInterval interval = new SimpleInterval(TEST_HEADER.getSequence(i).getSequenceName(), start, end);
+            final SimpleInterval paddedInterval = interval.expandWithinContig(pad, TEST_HEADER.getSequenceDictionary());
+            reads.addAll(oneBaseReadsUntilPosition(i, paddedInterval.getEnd() + 1));
+            shards.add(new ShardBoundary(interval, paddedInterval));
+        }
+
+        // create shard iterator
+        // create a list of reads larger than the interval
+        final ShardingIterator<GATKRead> it = new ShardingIterator<>(
+                reads.iterator(),
+                shards,
+                TEST_HEADER.getSequenceDictionary());
+
+        for (final ShardBoundary s: shards) {
+            testNextElement(it, s.getInterval(), s.getPaddedInterval(), s.getPaddedInterval().getLengthOnReference());
```

Comment:
Also assert that you get the correct number of shards back from your iterator (applies to your other tests as well).

Reply:
I think that it is implicitly tested because I am iterating over the `ShardBoundary` list: if the number of elements returned is smaller it will fail because there is no element; if the number of elements is smaller, the next assertion of hasNext will fail.

Should I add explicitly a test for this by adding an `int` variable and assert that it is equal to the `shards.size()`?